File: nes Watch App/Emulator/MapperNROM.swift
import Foundation

final class MapperNROM: Mapper {
    private let prgBanks: Int
    private let chrBanks: Int

    init(prgBanks: Int, chrBanks: Int) {
        self.prgBanks = prgBanks
        self.chrBanks = chrBanks
    }

    func mapCpuRead(_ addr: UInt16) -> Int? {
        guard addr >= 0x8000 else { return nil }
        if prgBanks > 1 {
            return Int(addr & 0x7FFF)
        }
        return Int(addr & 0x3FFF)
    }

    func mapCpuWrite(_ addr: UInt16) -> Int? {
        guard addr >= 0x8000 else { return nil }
        if prgBanks > 1 {
            return Int(addr & 0x7FFF)
        }
        return Int(addr & 0x3FFF)
    }

    func mapPpuRead(_ addr: UInt16) -> Int? {
        guard addr < 0x2000 else { return nil }
        return Int(addr)
    }

    func mapPpuWrite(_ addr: UInt16) -> Int? {
        guard addr < 0x2000 else { return nil }
        if chrBanks == 0 {
            return Int(addr)
        }
        return nil
    }
}

File: nes Watch App/Core/src/mapper_nrom.c
#include "../include/mapper_nrom.h"

int mapper_nrom_cpu_map(const MapperNROM *mapper, uint16_t addr) {
    if (addr < 0x8000) {
        return -1;
    }
    if (mapper->prgBanks > 1) {
        return (int)(addr & 0x7FFF);
    }
    return (int)(addr & 0x3FFF);
}

int mapper_nrom_ppu_map(const MapperNROM *mapper, uint16_t addr, bool write) {
    if (addr >= 0x2000) {
        return -1;
    }
    if (write && mapper->chrBanks != 0) {
        return -1;
    }
    return (int)addr;
}
